sygnal/apnspushkin.py

```diff
@@ -351,7 +360,8 @@ def _get_payload_event_id_only(
         payload = {}
 
         if device.data:
-            payload.update(device.data.get("default_payload", {}))
+            default_payload = device.data.get("default_payload", {})
+            payload.update(default_payload)
```

Comment:
Does this use of `default_payload` need checking for not being a dict too? (Not sure of the bigger picture here)

Reply:
My thinking on this was that we do the check on `default_payload` several lines before the only call to the function where this code is located (`_get_payload_event_id_only`), and the check returns/stops execution if the default payload is not a dict, so afaict the execution should never get to this stage if `default_payload` is not a dict.

Reply:
I think your reasoning is correct; but I have in mind:

- maintainability: in the future, we might call `_get_payload_event_id_only` with a `device` whose `default_payload` hasn't been validated
- ease of reading: it'd be nice to arrange this in a way that the reader doesn't need to repeat this reasoning every time we come to look at `default_payload`.

(Warning: the following is splitting hairs.)

Looking at the body of `_get_payload_event_id_only`, I note that the only time we use the `device` argument is to retrieve the `default_payload` (see also https://github.com/matrix-org/sygnal/pull/127). I'm tempted to suggest:

- remove `_get_payload_event_id_only`'s `device` argument
- have it take a `default_payload: Dict[str, Any]` instead
- change the caller to pass `default_payload` directly, handling the case where `device.data is None`

Reply:
Great I will do that, thanks for outlining your reasoning!

Reply:
(To emphasise: this is fine as it is and I _am_ going a little too extreme here. I think this is great as written and definitely mergeable. But I wanted to outline the thinking behind the question.)

Reply:
Ahh, I had something slightly different in mind. I was thinking:

- above line 288 add `default_payload = {}`
- this means we can remove lines 297 through 301, and always use 302 (because we _always_ have a default payload)

Reply:
Right that makes more sense, have updated.